### PR TITLE
Add XRPL (Ripple) support

### PR DESCRIPTION
**Pull request by jzethar**
This PR contains modules for Ripple blockchain: main module, NFT and token modules

### CODE REVIEW

**Comment by Har01d** [comment]
The main module is being tested now: https://3xpl.com/xrp-ledger

File: Modules/Common/RippleLikeFTModule.php [hunk 1]
```diff
@@ -0,0 +1,607 @@
...
+                        $flag_assets = 0;
+                        if ($broker_op)
+                        {
+                            if (is_array($tx['NFTokenBrokerFee'])) 
+                            {
+                                $currency = $tx['NFTokenBrokerFee']['currency']; 
+                                $broker_fee = $this->to_96($tx['NFTokenBrokerFee']['value']);
+                                $issuer = $tx['NFTokenBrokerFee']['issuer'];
+                                $flag_assets++;
+                            } else {
```
Comment (alexqrid) [inline Modules/Common/RippleLikeFTModule.php:277]: @jzethar fully comment this useless else please

File: Modules/Common/RippleLikeFTModule.php [hunk 2]
```diff
@@ -0,0 +1,607 @@
...
+                                        $new_owner = $tx['Account'];
+                                        if (is_array($affection['DeletedNode']['FinalFields']['Amount'])) {
+                                            if((!is_null($currency) && !is_null($issuer)) && ($currency . '.' . $issuer) !== $affection['DeletedNode']['FinalFields']['Amount']['currency'] . '.' .
+                                                             $affection['DeletedNode']['FinalFields']['Amount']['issuer'])
+                                            {
+                                                throw new ModuleError("Change of currency in NFTokenAcceptOffer");
+                                            }
+                                            $pay = $this->to_96($affection['DeletedNode']['FinalFields']['Amount']['value']);
+                                            $flag_assets++;
+                                        } else {
```
Comment (alexqrid) [inline Modules/Common/RippleLikeFTModule.php:299]: same as previous else, have to be commented.

File: Modules/Common/RippleLikeFTModule.php [hunk 3]
```diff
@@ -0,0 +1,607 @@
...
+                                                if((!is_null($currency) && !is_null($issuer)) && 
+                                                    ($currency . '.' . $issuer) !== 
+                                                    $affection['DeletedNode']['FinalFields']['Amount']['currency'] . '.' . 
+                                                                 $affection['DeletedNode']['FinalFields']['Amount']['issuer'])
+                                                {
+                                                    throw new ModuleError("Change of currency in NFTokenAcceptOffer");
+                                                }
+                                                $pay = $this->to_96($affection['DeletedNode']['FinalFields']['Amount']['value']);
+                                                $flag_assets++;
+                                            } else {
```
Comment (alexqrid) [inline Modules/Common/RippleLikeFTModule.php:321]: comment

File: Modules/Common/RippleLikeFTModule.php [hunk 4]
```diff
@@ -0,0 +1,607 @@
...
+                                                if((!is_null($currency) && !is_null($issuer)) && 
+                                                    ($currency . '.' . $issuer) !== 
+                                                    $affection['DeletedNode']['FinalFields']['Amount']['currency'] . '.' .
+                                                                 $affection['DeletedNode']['FinalFields']['Amount']['issuer'])
+                                                {
+                                                    throw new ModuleError("Change of currency in NFTokenAcceptOffer");
+                                                }
+                                                $prev_pay = $this->to_96($affection['DeletedNode']['FinalFields']['Amount']['value']);
+                                                $flag_assets++;
+                                            } else {
```
Comment (alexqrid) [inline Modules/Common/RippleLikeFTModule.php:339]: comment

File: Modules/Common/RippleTraits.php [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+                $num = substr($number, 0, $point) . substr($number, $point + 1);
+            else
+                $num = $number;
+            $pad = str_pad('', $zeros_in_the_end, '0');
+            $num .= $pad;
+            $num = ltrim($num, '0');
+            return $num;
+        }
+    }
+}
```
Comment (alexqrid) [inline Modules/Common/RippleTraits.php:213]: add new line to the end of file please

File: Modules/RippleMainModule.php [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+        ];
+    }
+}
```
Comment (alexqrid) [inline Modules/RippleMainModule.php:26]: no new line

**Review by alexqrid — CHANGES_REQUESTED** [review]
Need some minor style and logic (extra `else`) fixes

**Comment by Har01d** [comment]
Merging this despite it does not support the latest transaction types (the recent XRPL upgrade).

**Review by Har01d — APPROVED** [review]

**Pull request merged**